# Admixture Mapping
Admixture mapping is the process of assigning loci within the genome of an admixed individual to parental populations using allele frequencies in a phased dataset.

Overview of the method: [Winkler et al. 2010](https://doi.org/10.1146/annurev-genom-082509-141523)


## Phase Alleles
### Remove variants with missing genotypes
```bash
# Performed in UFRC queue system. See filter_missing.job for more details.
# Resources: 75 Mb, 1 min

module load vcftools/0.1.16
module load bcftools/1.15

INDIR="/blue/soltis/kasey.pham/euc_hyb_reseq/call_snps/04.filter_snps/maf0.05"
NAME="meehan_all_fil_maf0.05_snps"

export MISS=1.0

cat "$INDIR"/"$NAME".vcf | vcftools --vcf - --max-missing $MISS --recode --stdout | bcftools sort -O v - > "$NAME"_nomiss.vcf
```

### Filter all variants with samples with haploid GT calls and remove unanchored contigs
```bash
LIST_DIR="/blue/soltis/kasey.pham/euc_hyb_reseq"
module load bcftools/1.15

cat meehan_all_fil_maf0.05_snps_nomiss.vcf | bcftools view -S "$LIST_DIR"/sample_ids.txt | bcftools filter -e 'FORMAT/GT="hap"' -O v -o meehan_all_fil_maf0.05_snps_nohap.vcf 
```

### Phase alleles with [`Beagle`](https://faculty.washington.edu/browning/beagle/beagle.html)

```bash
# Performed in UFRC queue system. See filter_missing.job for more details.
# Resources: 1.52 Mb, 30s
module load beagle/5.2

WDIR="/blue/soltis/kasey.pham/euc_hyb_reseq/05.analyses/admix_map"
NAME="all_to_ASM1654582_fil_maf0.05_snps_nohap"

"$HPC_BEAGLE_BIN"/beagle gt="$NAME".vcf out="$NAME"_phased impute=false burnin=5 iterations=20 phase-states=280 nthreads=11
```

### Remove unanchored contigs from VCF

```bash
module load vcftools/0.1.16

vcftools --gzvcf meehan_all_fil_maf0.05_snps_nohap_phased.vcf.gz --recode --not-chr ChrUn --out meehan_all_fil_maf0.05_snps_nohap_phased_noChrUn.vcf
```

### Curate genetic map
I used a genetic map of DArT markers from a previous study (Butler et al. 2012). Physical locations for DArT markers were generated by Jakob Butler by BLASTing to _E. globulus_ reference genome and taking the best hit. Individual entries were therefore examined and manually curated to a reduced map of markers which were colinear in genetic and physical distance, with one marker per base pair site. Markers were selected or omitted based on the following criteria from most to least priority:

1. Maximizing the amount of markers which could be included.
2. By largest sequence identity to reference genome.
3. Markers with single mapping locations over multiple mapping locations.

All markers mapped to the aggregate chromosome for unplaced scaffolds or unmapped markers were removed.

Of the 893 markers mapped to chromosomes 1 through 11, 459 were retained.

The raw map can be viewed at [1060_LH_F2.xlsx](). The manually filtered map can be viewed at [1060_LH_F2_manual_copy.xlsx](). Extra columns were added to record markers flagged for removal or for location substitution.

Remove flagged columns and export genetic map in PLINK format in `R`:

```R
# Import genetic map file as dataframe and format
map <- read.csv("1060_LH_F2_manual_copy.csv", header = TRUE, na.strings = c("#N/A"))
map_colnames <- c("Marker", "LG", "cM", "Seq", "Chr_ANU_glob", "bp_ANU_glob", "Chr_ANU_grand", "bp_ANU_grand", "Chr_X46_glob", "bp_X46_glob", "delete_dupes", "delete_order")
colnames(map) <- map_colnames

# Remove flagged markers
map_final_bp <- map$bp_X46_glob
removed_markers <- which(map$delete_dupes == "x" | map$delete_order == "x") # 434
map_final_bp <- map_final_bp[-removed_markers]

# Generate the rest of the columns for PLINK genetic map and remove flagged markers
map_final_chr <- map$Chr_X46_glob
map_final_chr <- map_final_chr[-removed_markers]

map_final_name <- map$Marker
map_final_name <- map_final_name[-removed_markers]

map_final_cm <- map$cM
map_final_cm <- map_final_cm[-removed_markers]

map_final <- as.data.frame(cbind("Chr" = map_final_chr, "Marker" = map_final_name, "cM" = map_final_cm, "bp" = map_final_bp)) # 459 markers total
map_final$bp <- as.numeric(map_final$bp)

# Check markers mapped to the same place
dupes <- which(duplicated(map_final$bp))

# Check for out-of-order markers
# Instantiate loop storage variables
prev_bp <- map_final[1,"bp"]
prev_chr <- map_final[1, "Chr"]
out_of_order <- c()
# Loop through every row in map where there wasn't a duplicate marker mapping to same bp
for(i in seq(2,nrow(map_final))){
    # Examine current row
    curr_bp <- map_final[i, "bp"]
    curr_chr <- map_final[i, "Chr"]
    # If on the same chromosome...
    if(curr_chr == prev_chr){
        # Compare physical mapping of current marker to previous marker
        if(curr_bp < prev_bp){
            # If physical mapping is less than the previous marker, record in storage vector and update current bp position
            out_of_order <- c(out_of_order, i)
            prev_bp <- curr_bp
        } else {
            # If physical mapping is greater than or equal to previous marker, it's fine. Just update current bp position
            prev_bp <- curr_bp
        }
    # If on different chromosomes...
    } else {
        # Update the current chromosome and bp position to the new start, then continue with loop to compare to this baseline.
        prev_chr <- curr_chr
        prev_bp <- curr_bp
    }
}

# Print PLINK format genetic map
write.table(map_final, file = "1060_LH_F2_manual_copy.map", row.names = FALSE, col.names = FALSE, quote = FALSE, sep = "\t")
```

### Run admixture mapping software [`flare`](https://github.com/browning-lab/flare)

```bash

```