# Plot output of `bcftools query` for VCF annotations
# Usage: Rscript plot_snp_stats.r [depth_file] [gq_file] [sp_file] [info_file] [out_prefix]
# Note: assumes all outputs of BCFTools query have headers, which are not generated by default by the program.

dp_filename <- commandArgs(trailingOnly = TRUE)[1]
gq_filename <- commandArgs(trailingOnly = TRUE)[2]
sp_filename <- commandArgs(trailingOnly = TRUE)[3]
info_filename <- commandArgs(trailingOnly = TRUE)[4]
out_prefix <- commandArgs(trailingOnly = TRUE)[5]

dp <- read.table(dp_filename, header = TRUE, na.strings=".")
total_dp <- apply(dp, 1, sum)
rm(dp)

png(filename = paste(out_prefix, "dp_distr.png", sep = "_"),
    width = 1000, height = 1000, unit = "px")
    hist(total_dp, breaks = 50, main = paste(out_prefix, "total depth distribution"), xlab = "Total DP")
dev.off()

gq <- read.table(gq_filename, header = TRUE, na.strings=".")
avg_gq <- apply(gq, 1, mean)
rm(gq)

png(filename = paste(out_prefix, "gq_distr.png", sep = "_"),
    width = 1000, height = 1000, unit = "px")
    hist(avg_gq, breaks = 50, main = paste(out_prefix, "average genotype quality distribution"), xlab = "Average GQ")
dev.off()

png(filename = paste(out_prefix, "dp_vs_gq.png", sep = "_"),
    width = 1000, height = 1000, unit = "px")
    plot(x = avg_gq, y = total_dp, main = paste(out_prefix, "depth/genotype quality correlation"), xlab = "GQ", ylab = "DP")
dev.off()

rm(avg_gq)

sp <- read.table(sp_filename, header = TRUE, na.strings=".")
avg_sp <- apply(sp, 1, mean)
rm(sp)

png(filename = paste(out_prefix, "sp_distr.png", sep = "_"),
    width = 1000, height = 1000, unit = "px")
    hist(avg_sp, breaks = 50, main = paste(out_prefix, "average strand bias distribution"), xlab = "Average SP")
dev.off() 

rm(avg_sp)

info <- read.table(info_filename, header = TRUE, na.strings=".", colClasses = c("character", "numeric", "numeric", "numeric", "numeric", "numeric"))

png(filename = paste(out_prefix, "qual_distr.png", sep = "_"),
    width = 1000, height = 1000, unit = "px")
    hist(info$QUAL, breaks = 50, main = paste(out_prefix, "variant quality distribution"), xlab = "QUAL")
dev.off()

png(filename = paste(out_prefix, "dp_vs_qual.png", sep = "_"),
    width = 1000, height = 1000, unit = "px")
    plot(x = info$QUAL, y = total_dp, main = paste(out_prefix, "depth/quality correlation"), xlab = "QUAL", ylab = "DP")
dev.off()

write(paste("done plotting", out_prefix), stdout())